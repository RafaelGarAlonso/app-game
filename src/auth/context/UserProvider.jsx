import { UserContext } from './UserContext';
import { INIT_USER, INIT_USERS_TABLE } from '../../constants/users'

export const UserProvider = ({ children }) => {

  const getUser = () => {
    const localStorageData = localStorage.getItem('currentUser');
    return localStorageData ? JSON.parse(localStorageData) : INIT_USER;
  };

  const setNewUserSession = ({ userName }) => {
    const listOfUsers = getListOfUsers();

    for (const [index, user] of listOfUsers.entries()) {
      if (user.userName === userName) {
        const { userName, points } = listOfUsers[index];
        localStorage.setItem('currentUser', JSON.stringify({ userName, points }));
        return false;
      }
    };

    localStorage.setItem('currentUser', JSON.stringify({ userName, points: 0 }));
    return true;
  };

  const getAppLoaded = () => {
    return sessionStorage.getItem('appLoaded');
  };

  const setAppLoaded = (loaded) => {
    sessionStorage.setItem('appLoaded', loaded);
  };

  const userEarnPoints = () => {
    const user = getUser();
    localStorage.setItem('currentUser', JSON.stringify({ userName: user.userName, points: user.points + 1 }));
  };

  const getListOfUsers = () => {
    const listOfUsers = localStorage.getItem('listOfUsers');
    return listOfUsers ? JSON.parse(listOfUsers) : null;
  };

  const setNewListOfUsers = (listOfUsers = INIT_USERS_TABLE) => {
    localStorage.setItem('listOfUsers', JSON.stringify(listOfUsers));
  };

  const setPointsToListUsers = (currentUser) => {
    const listOfUsers = getListOfUsers();

    for (let [index, user] of listOfUsers.entries()) {
      if (user.userName === currentUser.userName) {
        listOfUsers[index].points = currentUser.points;
        break;
      }
    };

    setNewListOfUsers(listOfUsers);
  };

  return (
    <UserContext.Provider value={{
      setNewUserSession,
      getUser,
      setAppLoaded,
      getAppLoaded,
      userEarnPoints,
      getListOfUsers,
      setNewListOfUsers,
      setPointsToListUsers
    }}>
      {children}
    </UserContext.Provider>
  )
}