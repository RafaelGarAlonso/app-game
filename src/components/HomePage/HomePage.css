.container-home {
  width: 100%;
  position: absolute;
  top: 40%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.button-login {
  min-width: 10rem;
  min-height: 3rem;
  justify-content: center;
}

.button-register {
  display: flex; 
  align-items: center;
}

.button-register button {
  max-width: 50%;
  padding: 0.2rem;
}

.form {
  flex-direction: column;
  width: 100%;
  display: flex;
}

.login-form {
  justify-content: center;
}

.login-user {
  display: flex;
  width: 50%;
  justify-content:center;
  margin: 40px auto 40px auto;
  border: 1px dotted #747474;
}

.profile {
  text-align: center;
}

.button-score-table {
  position: absolute;
  width: 150px;
  margin: 20px;
  bottom: 0;
  right: 0;
}