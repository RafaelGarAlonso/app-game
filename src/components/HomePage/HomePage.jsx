import { Container, Stack, Form, Button, Image } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import { useState, useContext, useEffect } from 'react';
import { UserContext } from '../../auth/context/UserContext';
import './HomePage.css';

export const HomePage = () => {
    const [form, setForm] = useState({});
    const [errors, setErrors] = useState({});
    const { setNewUserSession, getListOfUsers, setNewListOfUsers } = useContext(UserContext);
    const navigate = useNavigate();

    useEffect(() => {
        if (!getListOfUsers()) setNewListOfUsers();
    }, [getListOfUsers, setNewListOfUsers]);

    const onFormSubmit = (e) => {
        e.preventDefault();
        const newErrors = findFormErrors();
        if (Object.keys(newErrors).length > 0) {
            setErrors(newErrors);
        } else {
            const newUser = { userName: form.username, points: 0 }
            const isNewUser = setNewUserSession(newUser);

            if (isNewUser) setNewListOfUsers(getListOfUsers().concat(newUser));

            setTimeout(() => {
                navigate('/game');
            }, 25);
        }
    };

    const findFormErrors = () => {
        const { username } = form;
        const newErrors = {};
        if (!username || username.trim() === '') newErrors.username = 'Please enter a valid name';
        return newErrors;
    };

    const setField = (field, value) => {
        setForm({ ...form, [field]: value });
        if (errors[field]) {
            setErrors({ ...errors, [field]: null })
        };
    };

    const loadScoreTable = () => {
        navigate('/score-table');
    }

    return (
        <>
            <Container className="container-home" fluid>
                <h1 className="profile">Create a profile or resume your game</h1>
                <Image className="login-user" src="assets/user.png" roundedCircle />
                <Stack gap={2} className="col-md-5 mx-auto m-4">
                    <Form className="form" onSubmit={onFormSubmit}>
                        <Form.Floating className="mb-3">
                            <Form.Control
                                className="form-input"
                                onChange={e => setField('username', e.target.value)}
                                type="text"
                                maxLength={12}
                                placeholder="Name"
                                isInvalid={!!errors.username}
                            />
                            <Form.Control.Feedback type="invalid">
                                {errors.username}
                            </Form.Control.Feedback>
                            <label htmlFor="floatingInputCustom">Enter a name</label>
                        </Form.Floating>
                        <Button className="form-button" variant="primary" type="submit">
                            Start
                        </Button>
                    </Form>
                </Stack>

            </Container>
            <Button onClick={loadScoreTable} className="button-score-table" size="lg" variant="info">Score Table</Button>
        </>
    )
}