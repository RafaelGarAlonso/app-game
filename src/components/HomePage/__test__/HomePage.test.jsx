import { render, fireEvent, screen } from '@testing-library/react';
import { HomePage } from '../HomePage';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../../auth/context/UserProvider';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom')),
    useNavigate: () => mockedUsedNavigate,
}));

describe('Test <HomePage />', () => {
    const renderTemplate = () => {
        return render(
            <UserProvider>
                <MemoryRouter>
                    <HomePage />
                </MemoryRouter>
            </UserProvider>
        );
    }

    test('Check snapshot component page', () => {
        const { container } = renderTemplate();
        expect(container).toMatchSnapshot();
    });

    test('It should click on button score table and navigate to /score-table', () => {
        const { container } = renderTemplate();
        const scoreTableButton = container.querySelector(".button-score-table");

        fireEvent.click(scoreTableButton);
        expect(mockedUsedNavigate).toBeCalled();
    });

    test('It should click on form when is a invalid name', () => {
        const { container } = renderTemplate();
        const formButton = container.querySelector(".form-button");

        fireEvent.click(formButton);
        const fieldInvalid = container.querySelector(".invalid-feedback");
        expect(fieldInvalid.innerHTML).toBe("Please enter a valid name");
    });

    test('It should click on form when is a valid name', () => {
        const { container } = renderTemplate();

        const input = screen.getByPlaceholderText("Name");
        const formButton = container.querySelector(".form-button");
        const value = "Manolo";

        fireEvent.change(input, { target: { value } });
        fireEvent.click(formButton);

        // TODO: Spy metho onFormSubmit to confirm callback code executed
    });
});