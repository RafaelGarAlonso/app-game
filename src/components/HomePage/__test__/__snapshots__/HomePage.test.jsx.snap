// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <HomePage /> Check snapshot component page 1`] = `
<div>
  <div
    class="container-home container-fluid"
  >
    <h1
      class="profile"
    >
      Create a profile or resume your game
    </h1>
    <img
      class="login-user rounded-circle"
      src="assets/user.png"
    />
    <div
      class="col-md-5 mx-auto m-4 vstack gap-2"
    >
      <form
        class="form"
      >
        <div
          class="mb-3 form-floating"
        >
          <input
            class="form-input form-control"
            maxlength="12"
            placeholder="Name"
            type="text"
          />
          <div
            class="invalid-feedback"
          />
          <label
            for="floatingInputCustom"
          >
            Enter a name
          </label>
        </div>
        <button
          class="form-button btn btn-primary"
          type="submit"
        >
          Start
        </button>
      </form>
    </div>
  </div>
  <button
    class="button-score-table btn btn-info btn-lg"
    type="button"
  >
    Score Table
  </button>
</div>
`;
